# World Happiness Report - Exploratory Practice

**Dataset:** World Happiness Report (source: https://www.kaggle.com/datasets/unsdsn/world-happiness/data)

**Goal:** Practice pandas fundamentals (inspecting, filtering, sorting, grouping, merging) by exploring what factors relate to happiness scores across countries.

**Date started:** 2026-09-04

## 1. Imports & Setup

In [117]:
import pandas as pd 
import numpy as np

## 2. Loading Data

In [118]:
df_2019 = pd.read_csv('data/2019.csv')

## 3. Inspecting Data

**Q: What are the shape, column names, and dtypes of the dataframe?**

In [119]:
df_2019.shape

(156, 9)

In [120]:
df_2019.info()

<class 'pandas.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    str    
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), str(1)
memory usage: 11.1 KB


In [121]:
df_2019.head()

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298


In [122]:
df_2019.tail()

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035
155,156,South Sudan,2.853,0.306,0.575,0.295,0.010,0.202,0.091


**Dataset context:**
* There are 156 rows, each representing a country in the dataset, and 9 columns. 
* The first 3 columns are the country's rank, the country, and its happiness score.
* The next 6 columns describe the extent to which each of 6 factors - GDP, social support, life expectancy, freedom, generosity, and absence of corruption - contribute to making life evaluations higher in each country than they are in Dystopia. 
* Dystopia is a hypothetical country that has values equal to the world's lowest national averages for each of the 6 factors. It is a benchmark against which other countries are compared. 

**Q: Are there are any misisng values, and in which columns?**

In [123]:
df_2019.sample(10)

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
136,137,Egypt,4.166,0.913,1.039,0.644,0.241,0.076,0.067
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147
105,106,South Africa,4.722,0.960,1.351,0.469,0.389,0.130,0.055
98,99,Ivory Coast,4.944,0.569,0.808,0.232,0.352,0.154,0.090
122,123,Mozambique,4.466,0.204,0.986,0.390,0.494,0.197,0.138
93,94,Vietnam,5.175,0.741,1.346,0.851,0.543,0.147,0.073
19,20,Czech Republic,6.852,1.269,1.487,0.920,0.457,0.046,0.036
132,133,Ukraine,4.332,0.820,1.390,0.739,0.178,0.187,0.010
107,108,Venezuela,4.707,0.960,1.427,0.805,0.154,0.064,0.047
14,15,United Kingdom,7.054,1.333,1.538,0.996,0.450,0.348,0.278


In [124]:
# checks if there are NaN values in any of the columns
df_2019.isna().any()

Overall rank                    False
Country or region               False
Score                           False
GDP per capita                  False
Social support                  False
Healthy life expectancy         False
Freedom to make life choices    False
Generosity                      False
Perceptions of corruption       False
dtype: bool

* `.isna()` returns a dataframe the same shape as `df_2019`, filled with `True`/`False`:
    * `True` wherever the original dataframe has a missing value (`NaN`)
    * `False` everywhere else

* `.any()` returns whether any element is `True` in a series or dataframe

There does not seem to be any missing values. The dataset looks well cleaned.

**Q: What is the mean, min, and max of the happiness score column?**

In [125]:
df_2019['Score'].describe()

count    156.000000
mean       5.407096
std        1.113120
min        2.853000
25%        4.544500
50%        5.379500
75%        6.184500
max        7.769000
Name: Score, dtype: float64

In [126]:
df_2019['Score'].mean()

np.float64(5.407096153846155)

In [127]:
df_2019['Score'].min()

np.float64(2.853)

In [128]:
df_2019['Score'].max()

np.float64(7.769)

* Mean happiness score: 5.407
* Min happiness score: 2.853
* Max happiness score: 7.769

## 4. Cleaning Data

In [129]:
df_2019

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298
...,...,...,...,...,...,...,...,...,...
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035


Rename column headers to all lowercase and without spaces, so they can be more conveniently accessed later.

In [130]:
mapping = {
    'Overall rank' : 'overall_rank',
    'Country or region' : 'country',
    'Score' : 'score',
    'GDP per capita' : 'gdp_per_capita',
    'Social support' : 'social_support',
    'Healthy life expectancy' : 'life_expectancy',
    'Freedom to make life choices' : 'freedom',
    'Generosity' : 'generosity',
    'Perceptions of corruption' : 'perceived_corruption'
}

df_2019 = df_2019.rename(columns=mapping)

df_2019.head()

,overall_rank,country,score,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298


## 5. Analysis: Sorting & Filtering

**Q: Create a dataframe of the 10 happiest countries, keeping only the rank, country name, and score columns**

In [131]:
# get top 10 ranked countries using boolean indexing
df_top_10_countries = df_2019[df_2019['overall_rank'] <= 10]

# alternatively: df_2019.head(10)

# keep only desired columns by indexing with list of column names
df_top_10_countries = df_top_10_countries[['overall_rank', 'country', 'score']]

# reset index to start from 0, and drop old index column
df_top_10_countries.reset_index(drop=True, inplace=True)

df_top_10_countries

,overall_rank,country,score
0,1,Finland,7.769
1,2,Denmark,7.600
2,3,Norway,7.554
3,4,Iceland,7.494
4,5,Netherlands,7.488
5,6,Switzerland,7.480
6,7,Sweden,7.343
7,8,New Zealand,7.307
8,9,Canada,7.278
9,10,Austria,7.246


**Q: Create a dataframe of the 10 least happy countries.**

In [132]:
# get lowest 10 ranked countries
df_lowest_10_countries = df_2019.tail(10)

# keep only desired columns by indexing with list of column names
df_lowest_10_countries = df_lowest_10_countries[['overall_rank', 'country', 'score']]

# reset index to start from 0, drop old index colum
df_lowest_10_countries.reset_index(drop=True, inplace=True)

df_lowest_10_countries

,overall_rank,country,score
0,147,Haiti,3.597
1,148,Botswana,3.488
2,149,Syria,3.462
3,150,Malawi,3.410
4,151,Yemen,3.380
5,152,Rwanda,3.334
6,153,Tanzania,3.231
7,154,Afghanistan,3.203
8,155,Central African Republic,3.083
9,156,South Sudan,2.853


**Q: Among countries with GDP per capita above the median, which 10 have the lowest happiness score?**

In [133]:
# calculate median GDP per capita
median_gdp = df_2019['gdp_per_capita'].median()

# get countries with GDP per capita above the median
df_gdp_above_median = df_2019[df_2019['gdp_per_capita'] > median_gdp]

# among those, get 10 countries with lowest happiness score
df_gdp_above_median.nsmallest(10, 'score')

# alternatively: df_gdp_above_median.tail(10)

,overall_rank,country,score,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption
147,148,Botswana,3.488,1.041,1.145,0.538,0.455,0.025,0.100
125,126,Iraq,4.437,1.043,0.980,0.574,0.241,0.148,0.089
116,117,Iran,4.548,1.100,0.842,0.785,0.305,0.270,0.125
103,104,Gabon,4.799,1.057,1.183,0.571,0.295,0.043,0.055
96,97,Bulgaria,5.011,1.092,1.513,0.815,0.311,0.081,0.004
92,93,China,5.191,1.029,1.125,0.893,0.521,0.058,0.100
90,91,Lebanon,5.197,0.987,1.224,0.815,0.216,0.166,0.027
89,90,Azerbaijan,5.208,1.043,1.147,0.769,0.351,0.035,0.182
87,88,Algeria,5.211,1.002,1.160,0.785,0.086,0.073,0.114
86,87,Turkmenistan,5.247,1.052,1.538,0.657,0.394,0.244,0.028


`nsmallest()` returns the `n` rows with the smallest values from a dataframe or series
* Syntax: `df.nsmallest(n, columns, keep)`
* Parameters: 
    * `n`: number of items to retrieve
    * `columns`: column name, or list of column names
    * `keep`: specifies what to do with duplicate rows

## 6. Analysis: Grouping

**Q: Split countries into low, medium, and high "tiers" based on the GDP per capita score contribution (hint: using pd.cut or pd.qcut to create the buckets). What's the average happiness score in each tier?**

In [134]:
df_2019['gdp_per_capita'].describe()

count    156.000000
mean       0.905147
std        0.398389
min        0.000000
25%        0.602750
50%        0.960000
75%        1.232500
max        1.684000
Name: gdp_per_capita, dtype: float64

In [135]:
# create labels for tier categories
labels = ['Low', 'Medium', 'High']

# get a subset of df_2019 containing this list of columns
df_countries_by_gdp = df_2019[['overall_rank', 'country', 'score', 'gdp_per_capita']]

# divide gdp values into 3 equally sized tiers (low, medium, high)
# add a new column for these tiers
df_countries_by_gdp['gdp_tier'] = pd.qcut(df_2019['gdp_per_capita'], q=3, labels=labels)

df_countries_by_gdp

,overall_rank,country,score,gdp_per_capita,gdp_tier
0,1,Finland,7.769,1.340,High
1,2,Denmark,7.600,1.383,High
2,3,Norway,7.554,1.488,High
3,4,Iceland,7.494,1.380,High
4,5,Netherlands,7.488,1.396,High
...,...,...,...,...,...
151,152,Rwanda,3.334,0.359,Low
152,153,Tanzania,3.231,0.476,Low
153,154,Afghanistan,3.203,0.350,Low
154,155,Central African Republic,3.083,0.026,Low


In [136]:
# group the countries by the tier values
# for each group (low, medium, high), calculate the mean happiness score
df_countries_by_gdp.groupby('gdp_tier')['score'].mean() # returns a Series

gdp_tier
Low       4.428077
Medium    5.319365
High      6.473846
Name: score, dtype: float64

In [137]:
# convert series returned from groupby and aggregation into a dataframe
df_avg_score_by_gdp = df_countries_by_gdp.groupby('gdp_tier')['score'].mean().reset_index()
df_avg_score_by_gdp

,gdp_tier,score
0,Low,4.428077
1,Medium,5.319365
2,High,6.473846


* Countries were split into 3 equally-sized tiers based on GDP per capita. This was done using the `pd.qcut()` function, which splits a Series into approximately equally sized bins.
* A dataframe with only the `overall_rank`, `country`, `score`, and `gdp_per_capita` columns was obtained. A new `gdp_tier` column was added to display the GDP tier of each country.
* The countries were grouped by their `gdp_tier` column values. For each tier group (low, medium, high), the average happiness score was calculated by applying the `.mean()` function on the `score` column.
* The Series returned from the `groupby()` and `mean()` operations was converted to a DataFrame by applying the `.reset_index()` function. The index labels (0, 1, 2...) were pulled into a new column, changing the data structure from 1D (Series) to 2D (DataFrame).

Result: 
* Mean score for low tier countries: 4.43
* Mean score for medium tier countries: 5.32
* Mean score for high tier countries: 6.47

**Q: Repeat the same steps for the other 5 factors. Store the results in their own DataFrames.**

In [138]:
labels = ['Low', 'Medium', 'High']
factors = ['gdp_per_capita', 'social_support', 'life_expectancy', 'freedom', 'generosity', 'perceived_corruption']
# created dictionary of dataframes
dfs_avg_score_by_tier = {}

# for each factor, dynamically create its dataframe of average score by tier
# store each created dataframe into a dictionary of dataframes
for factor in factors:
    # create dataframe
    dfs_avg_score_by_tier[f'df_avg_score_by_{factor}'] = df_2019.groupby(
        pd.qcut(df_2019[factor], q=3, labels=labels)
        )['score'].mean().reset_index()

# verify by accessing a dataframe in dataframes dictionary
dfs_avg_score_by_tier['df_avg_score_by_social_support'].head()

,social_support,score
0,Low,4.342596
1,Medium,5.441288
2,High,6.437404


* The goal was to create 6 DataFrames that show the average score by tier, one for each factor. To do so, I looped through each factor in a list of factors, and inserted the factor name where needed dynamically using f-strings. 
* The line that creates the DataFrame is complex-looking, but it's just the combined version of what was done previously for the GDP per capita, but with the added use of f-strings. 

*A note on the six factors:*

The six factors are variables used *after* obtaining the survey results to *explain* the variation in happiness scores between countries. The numerical score assigned to each factor shows how much that specific condition contributes - positively or negatively - to a country's happiness score, compared to the hypothetical baseline Dystopia. 

**Q: For each of the six factors, how big is the jump between the "Low" tier and "High" tier average happiness score? This might reveal which factor most distinguishes happy from unhappy countries.**

In [193]:
factors = ['gdp_per_capita', 'social_support', 'life_expectancy', 'freedom', 'generosity', 'perceived_corruption']

# initialize list of lists, containing rows, to be converted to dataframe
factor_gaps_list = []

# for each of the 6 factors, get the score gap between the high and low tier average scores
for factor in factors:
    df = dfs_avg_score_by_tier[f'df_avg_score_by_{factor}']
    high = df.iat[2, 1]
    low = df.iat[0, 1]
    # add this score gap data for this factor as a list, to be converted into a row
    factor_gaps_list.append([factor, high - low])

# convert list of lists, containing rows, into a dataframe; specify column names
# the dataframe displays, for each factor, the gap in happiness scores between high and low tiers
df_score_gaps = pd.DataFrame(factor_gaps_list, columns=['factor', 'score_gap']) \
                            .sort_values('score_gap', ascending=False) \
                            .reset_index(drop=True)

df_score_gaps

,factor,score_gap
0,life_expectancy,2.153749
1,social_support,2.094808
2,gdp_per_capita,2.045769
3,freedom,1.355115
4,perceived_corruption,0.603567
5,generosity,0.245684


* The goal was to create a DataFrame displaying, for each factor, the gap in average happiness scores between the high and low country tiers.
* A list of lists, `factor_gap_list`, was initialized to store the data - the rows - that would be converted to the DataFrame. Each row contains the factor and the score gap for that factor.
* Iterating through each of the 6 factors, the `df_average_score_by_{factor}` DataFrame for that factor was accessed from the list of DataFrames (created in the previous segment).
    * The score gap was calculated by subtracting the score of the high tier from that of the low tier.
    * The scores for each the high and low tiers were accessed using the `.iat[rowIndex, colIndex]` function, which is used to access or modify a single value by its index position. 
    * This row  - th factor and the score gap - was appended to the list of lists data variable.
* The data gathered was converted into a DataFrame. To do this, the `factor_gaps_list` list of lists was passed as the `data` parameter into the `pd.Dataframe(data)` function. A list of column names were passed into the `columns` parameter, because that information is not given in the `data` list of lists. 

Result:
* Life expectancy was the factor with the greatest score gap: 2.153.
* Social support: 2.094
* GDP per capita: 2.046
* Freedom: 1.355
* Perceived corruption: 0.604
* Generosity was the factor with the lowest score gap: 0.246. 

**Q: For each country, find which of the six factor columns has the largest value — creating a new categorical column like "dominant factor." Then groupby that and count how many countries are "led" by each factor. Are there a few factors that most countries are mainly driven by, or is it more mixed?**

In [139]:
# for each country, find the maximum value between the factor columns
factors = ['gdp_per_capita', 'social_support', 'life_expectancy', 'freedom', 'generosity', 'perceived_corruption']
df_2019['dominant_factor'] = df_2019[factors].idxmax(axis=1)

df_2019

,overall_rank,country,score,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption,dominant_factor
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393,social_support
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410,social_support
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341,social_support
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118,social_support
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298,social_support
...,...,...,...,...,...,...,...,...,...,...
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411,social_support
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147,social_support
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025,social_support
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035,generosity


The `idxmax()` method returns, by default, a Series with the index of the maximum value for each column. 
* By default, the method searches down rows (row-wise) to find index of the maximum value in a *column*. 
* By specifying `axis=1` (or `axis=columns`), the method searches across columns (column-wise) to find index of the maximum value in a  *row*.

In [148]:
# group countries by dominant factor
# count the number of countries led by each dominant factor
df_countries_per_dominant_factor = df_2019['dominant_factor'].value_counts().reset_index()

# add missing rows of dominant factors (with 0 number of countries)
missing_data = [
    {'dominant_factor': 'life_expectancy', 'count' : 0},
    {'dominant_factor': 'freedom', 'count' : 0},
    {'dominant_factor': 'perceived_corruption', 'count' : 0},
]
df_missing_data = pd.DataFrame(missing_data)

# concatenate missing rows dataframe with dominant factors dataframe
df_countries_per_dominant_factor = pd.concat([df_countries_per_dominant_factor, df_missing_data], ignore_index=True) \
                                     .sort_values('count', ascending=False) \
                                     .reset_index(drop=True)

# rename dominant_factor column values to be more readable
mapping = {'social_support' : 'Social support', 
           'gdp_per_capita' : 'GDP per capita',
           'generosity' : 'Generosity', 
           'life_expectancy' : 'Life expectancy',
           'freedom' : 'Freedom', 
           'perceived_corruption' : 'Perceived corruption'}
df_countries_per_dominant_factor['dominant_factor'] = df_countries_per_dominant_factor['dominant_factor'].replace(mapping)

df_countries_per_dominant_factor

,dominant_factor,count
0,Social support,137
1,GDP per capita,18
2,Generosity,1
3,Life expectancy,0
4,Freedom,0
5,Perceived corruption,0


* The 'dominant factor' is the factor with the highest score of the six for a country - the factor that contributed the most to increasing its happiness score. 
* The 'dominant factor' for each country was found by applying the `idxmax()` function, which returns a Series with the index of the maximum value for each row. This Series was used to create a new column, `dominant_factor`. 
* The number of countries for each dominant factor was counted, and this count was used to create a new DataFrame showing the number of countries in each `dominant_factor` group. 
    * This was done by applying the `value_counts()` function to count the occurences of each unique value in the `dominant_factors` column.
    * The `.reset_index()` function was used to return a new DataFrame.

* The `df_countries_per_dominant_factor` DataFrame did not include all factors, because some of them were not dominant factors for any of the countries. However, for completeness, I added the missing factors (with a `count` of 0) to the DataFrame.
    * This was done by creating a list of dictionaries for the rows I wanted to add, which was transformed into a Pandas DataFrame. Then, using `pd.concat()`, the DataFrame created for the new rows was merged with the `df_countries_per_dominant_factor` DataFrame from earlier.
    * The `dominant_factor` column values were renamed for readability. This was done by passing a dictionary mapping the old names to the new names into the `replace()` function, which was applied to the `dominant_factor` column of the DataFrame.

Result:
* Social support was the dominant factor for 137 countries.
* GDP per capita was the dominant factor for 18 countries.
* Generosity was the dominant factor for 1 country. 
* Life expectancy, freedom, and perceived corruption were not the dominant factors for any country. 


In [ ]:
# run .describe() on all six factor columns to compare min, max, mean, etc.
factors = ['gdp_per_capita', 'social_support', 'life_expectancy', 'freedom', 'generosity', 'perceived_corruption']
df_2019[factors].describe()

,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption
count,156.000000,156.000000,156.000000,156.000000,156.000000,156.000000
mean,0.905147,1.208814,0.725244,0.392571,0.184846,0.110603
std,0.398389,0.299191,0.242124,0.143289,0.095254,0.094538
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.602750,1.055750,0.547750,0.308000,0.108750,0.047000
50%,0.960000,1.271500,0.789000,0.417000,0.177500,0.085500
75%,1.232500,1.452500,0.881750,0.507250,0.248250,0.141250
max,1.684000,1.624000,1.141000,0.631000,0.566000,0.453000


It was striking to see social support as the 'leading' dominant factor, by such a large margin. So, to see more details on the social support score, in comparison with the scores of other factors, I ran the `.describe()` function on all six factor columns, showing their mean, std, min, max, etc. 

To better understand why social support has the greatest score for the majority of countries, I should look into the report's methodology - how were the numerical scores for each factor estimated? 

On the other hand, how much does this reveal about what the most *meaningful* driver of a country's happiness score is? 